crypto/dh/dh_check.c

```diff
@@ -67,18 +68,14 @@ int DH_check_params(const DH *dh, int *ret)
 
 /*-
  * Check that p is a safe prime and
- * if g is 2, 3 or 5, check that it is a suitable generator
- * where
- * for 2, p mod 24 == 11
- * for 3, p mod 12 == 5
- * for 5, p mod 10 == 3 or 7
- * should hold.
+ * g is a suitable generator.
```

Comment:
The order of any element in a group of size `2q` with `q` a prime, is `1, 2, q` or `2q`. Only `1` and `p-1` have the first two orders, all other values of `g` have order `q` if `g` is a quadratic residue mod `p` and `2q` otherwise.  We should generally prefer quadratic residues and order `q` as this avoids having `Z_2` as a subgroup, and "leaking bit 0" (do we really care about that???).

The choice of `2` as a generator is a sound default, because it makes key agreement computations more efficient.  All the FFDHE groups for TLS 1.3 have `g = 2` with `g` a quadratic residue of order `q`: https://tools.ietf.org/html/rfc7919#appendix-A

What this means for OpenSSL is that we should generally prefer `p = 7 mod 8` to make `2` a quadratic residue, and also `p = 2 mod 3` so that `q = (p-1)/2` stands a chance of being prime, which combines to `p = 23 mod 24` and `q = 11 mod 12`.

If we wanted `2` to be a *non-residue* so that `g` generates the whole order `2q` multiplicative group, we'd need `p = 3 mod 8`, and `p = 2 mod 3` as before, and so `p = 11 mod 24`.  So OpenSSL previously prefers `o(g) = 2q`, but that's less preferred than `o(g) = q`. So our default choice of prime should switch from `11 mod 24` to `23 mod 24`.

Reply:
right.